### PR TITLE
Dynamicly retrieve formkey_salt

### PR DESCRIPTION
**Pull request by TunMax**
Dynamically retrieving formkey_salt can make the program more robust.

### CODE REVIEW

**Comment by TunMax** [comment]
this commit can fix the following error.

WARNING:root:Server returned a status code of 400 while downloading https://poe.com/api/gql_POST. Retrying (1/10)...
WARNING:root:Server returned a status code of 400 while downloading https://poe.com/api/gql_POST. Retrying (2/10)...
WARNING:root:Server returned a status code of 400 while downloading https://poe.com/api/gql_POST. Retrying (3/10)...
...

File: poe-api/src/poe.py [hunk 1]
```diff
@@ -199,12 +199,10 @@ def get_device_id(self):
     return device_id
 
   def extract_formkey(self, html):
-    script_regex = r'<script>if\(.+\)throw new Error;(.+)</script>'
-    script_text = re.search(script_regex, html).group(1)
     key_regex = r'var .="([0-9a-f]+)",'
-    key_text = re.search(key_regex, script_text).group(1)
```
Comment (ading2210) [inline poe-api/src/poe.py:205]: The changes here aren't necessary anymore.

File: poe-api/src/poe.py [hunk 2]
```diff
@@ -221,9 +219,14 @@ def get_next_data(self, overwrite_vars=False):
     json_regex = r'<script id="__NEXT_DATA__" type="application\/json">(.+?)</script>'
     json_text = re.search(json_regex, r.text).group(1)
     next_data = json.loads(json_text)
+    key_regex = r'(https://[^<]*_app.*?\.js)'
+    key_url = re.search(key_regex, r.text).group(1)
+    s = request_with_retries(self.session.get, key_url)
```
Comment (ading2210) [inline poe-api/src/poe.py:224]: I'd prefer for this to be in its own function rather than lumped into `get_next_data`.

File: poe-api/src/poe.py [hunk 3]
```diff
@@ -221,9 +219,14 @@ def get_next_data(self, overwrite_vars=False):
     json_regex = r'<script id="__NEXT_DATA__" type="application\/json">(.+?)</script>'
     json_text = re.search(json_regex, r.text).group(1)
     next_data = json.loads(json_text)
+    key_regex = r'(https://[^<]*_app.*?\.js)'
+    key_url = re.search(key_regex, r.text).group(1)
```
Comment (ading2210) [inline poe-api/src/poe.py:223]: Wouldn't this just search through the first script it finds rather than all of them?

File: poe-api/src/poe.py [hunk 4]
```diff
@@ -221,9 +219,14 @@ def get_next_data(self, overwrite_vars=False):
...
+    s = request_with_retries(self.session.get, key_url)
+    key_regex = r'\+"([a-zA-z0-9]{17})"'
```
Comment (ading2210) [inline poe-api/src/poe.py:225]: This regex is way too broad. This is searching for *every* 17 character alphanumeric string, which may turn up false positives.

**Review by ading2210 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**